docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,170 @@
+
+# Memory allocation routines
+## What is the problem?
+FEX-Emu needs to allocate memory differently than regular applications. This problem happens because FEX runs both 32-bit and 64-bit guest
+applications in the same address space as FEX itself. When running 32-bit applications, FEX reserves up all memory above 4GB in order to correctly
+emulate the 32-bit address space. We then use that reserved space for FEX allocations, so we don't interrupt application's own allocations.
+
+### Why not just just replace the system allocator?
+JEMalloc and similar allocators allow your application to replace the system allocator. This is something that FEX used to do, but this has since been
+removed because 32-bit thunks need to allocate memory in the lower 4GB.
+
+## APIs which allocate memory that FEX needs to avoid
+Most C++ APIs allow you to replace their allocators, but some don't and we need to avoid those APIs. If FEX uses them then the 32-bit application
+running might run out of memory. This isn't an all encompassing list and we will add to it as our CI captures more problems.
+
+### `get_nprocs_conf`
+Use `FEXCore::CPUInfo::CalculateNumberOfCPUs` instead.
+
+### `getcwd`
+Don't use getcwd with a nullptr buffer. It will allocate memory behind our back and return a pointer that needs a free.
+
+### `strerror`
+This allocates and frees memory based on locale! Even with C local it'll attempt to free(0).
+FEX-Emu should avoid using this function and instead just return the number.
+If necessary FEX will provide its own routine for getting this string back.
+
+### `std::make_unique`
+Use `fextl::make_unique` instead.
+
+### `std::unique_ptr`
+Use `fextl::unique_ptr` instead.
+
+### `std::filesystem`
+This namespace is /highly/ likely to allocate memory behind our back.
+FEX should avoid using this API as much as possible.
+
+#### std::filesystem::path
+#### std::filesystem::path::is_relative
+Use `FHU::Filesystem::IsRelative` instead.
+
+#### std::filesystem::absolute
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::exists
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::Exists` instead.
+
+#### std::filesystem::canonical
+Always allocates memory.
+Use `realpath` instead.
+
+#### std::filesystem::path::lexically_normal
+Use `FHU::Filesystem::LexicallyNormal` instead.
+
+#### std::filesystem::create_directory
+#### std::filesystem::create_directories
+Creates a std::filesystem::path when passing in to it.
+Use `FHU::Filesystem::CreateDirectory` and `FHU::Filesystem::CreateDirectories` instead.
+
+#### std::filesystem::path::parent_path
+Use `FHU::Filesystem::ParentPath` instead.
+
+#### std::filesystem::path::filename
+Use `FHU::Filesystem::GetFilename` instead.
+
+#### std::filesystem::copy_file
+Use `FHU::Filesystem::CopyFile` instead.
+
+### `std::fstream`
+This API always allocates memory and should be avoided.
```

Comment:
But we still use `fstream` in plenty of places. Do we need to replace them in the future, or should this guideline be changed to something like "make sure the `const char*` constructor is used"?

Reply:
It's not the `std::fs::path` constructors that allocate memory. It allocates memory internally that we can't control.